src/main/java/tools/jackson/databind/ObjectReader.java

```
@@ -1320,6 +1375,187 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
+     * @param p JsonParser to read from (will not be closed by this method)
+     * @return Deserialized object
+     * @throws tools.jackson.databind.exc.DeferredBindingException if recoverable problems were collected
+     * @throws tools.jackson.databind.DatabindException if a non-recoverable error occurred
+     * @since 3.1
+     */
+    public <T> T readValueCollecting(JsonParser p) throws JacksonException {
+        _assertNotNull("p", p);
+
+        // CRITICAL: Allocate a FRESH bucket for THIS call (thread-safety)
+        List<tools.jackson.databind.exc.CollectedProblem> bucket = new ArrayList<>();
+
+        // Create per-call attributes with the fresh bucket
+        ContextAttributes perCallAttrs = _config.getAttributes()
+            .withPerCallAttribute(tools.jackson.databind.deser.CollectingProblemHandler.class, bucket);
+
+        // Create a temporary ObjectReader with per-call attributes using public API
+        ObjectReader perCallReader = this.with(perCallAttrs);
+
+        try {
+            // Delegate to the temporary reader's existing readValue method
+            T result = perCallReader.readValue(p);
+
+            // Check if any problems were collected
+            if (!bucket.isEmpty()) {
+                // Check if limit was reached - read from per-call config to honor overrides
+                Integer maxProblems = (Integer) perCallReader.getConfig().getAttributes()
+                    .getAttribute(tools.jackson.databind.deser.CollectingProblemHandler.ATTR_MAX_PROBLEMS);
+                boolean limitReached = (maxProblems != null &&
+                    bucket.size() >= maxProblems);
+
+                throw new tools.jackson.databind.exc.DeferredBindingException(p, bucket, limitReached);
+            }
+
+            return result;
+
+        } catch (tools.jackson.databind.exc.DeferredBindingException e) {
+            throw e; // Already properly formatted
+
+        } catch (DatabindException e) {
```

Is the idea to only catch `DatabindException`s and not other `JacksonException`s?

---

Yes, that is intentional. DatabindException represents deserialization/binding errors that we want to collect (missing properties, type mismatches, etc.). 

Other Jackson exceptions (like JsonParseException, StreamReadException) might pertain more to malformed JSON structure that would fail fast and not be collected.